things/service/src/main/java/org/eclipse/ditto/things/service/persistence/actors/strategies/commands/MigrateThingDefinitionStrategy.java

```diff
@@ -0,0 +1,365 @@
+
+            final boolean conditionMatches = evaluateCondition(existingThing, conditionExpression, dittoHeaders);
+
+            final JsonPointer resourcePointer = JsonFactory.newPointer(resourceKey.getResourcePath());
+            if (!conditionMatches && doesMigrationPayloadContainResourceKey(migrationPayload, resourcePointer)) {
+                adjustedPayloadBuilder.remove(resourcePointer);
+            }
+        }
+
+        return adjustedPayloadBuilder.build();
+    }
+
+    private static boolean doesMigrationPayloadContainResourceKey(final JsonObject migrationPayload,
+            final JsonPointer pointer) {
+        return migrationPayload.getValue(pointer).isPresent();
+    }
+
+    private boolean evaluateCondition(final Thing existingThing,
+            final String conditionExpression,
+            final DittoHeaders dittoHeaders) {
+        try {
+            final var criteria = QueryFilterCriteriaFactory
+                    .modelBased(RqlPredicateParser.getInstance())
+                    .filterCriteria(conditionExpression, dittoHeaders);
+
+            final var predicate = ThingPredicateVisitor.apply(criteria,
+                    PlaceholderFactory.newPlaceholderResolver(TIME_PLACEHOLDER, new Object()));
+
+            return predicate.test(existingThing);
+        } catch (final ParserException | IllegalArgumentException e) {
+            throw InvalidRqlExpressionException.newBuilder()
+                    .message(e.getMessage())
+                    .cause(e)
+                    .dittoHeaders(dittoHeaders)
+                    .build();
+        }
+    }
+
+    private CompletionStage<Thing> generateSkeleton(
+            final MigrateThingDefinition command,
+            final DittoHeaders dittoHeaders) {
+        return wotThingSkeletonGenerator.provideThingSkeletonForCreation(
+                        command.getEntityId(),
+                        ThingsModelFactory.newDefinition(command.getThingDefinitionUrl()),
+                        dittoHeaders
+                )
+                .thenApply(optionalSkeleton -> {
+                    Thing skeleton = optionalSkeleton.orElseThrow(() ->
+                            SkeletonGenerationFailedException.newBuilder(command.getEntityId())
+                                    .dittoHeaders(command.getDittoHeaders())
+                                    .build()
+                    );
+
+                    return skeleton.toBuilder()
+                            .setDefinition(ThingsModelFactory.newDefinition(command.getThingDefinitionUrl()))
+                            .build();
+                });
+    }
+
+
+
+    private Thing extractDefinitions(final Thing thing) {
+        var thingBuilder = ThingsModelFactory.newThingBuilder();
+        thing.getFeatures().orElseGet(ThingsModelFactory::emptyFeatures).forEach(feature ->
+                thingBuilder.setFeature(feature.getId(), feature.getDefinition().get(), null));
```

Comment:
`feature.getDefinition().get()` might throw exception when the feature has no definition

also, passing `null` as last param is now allowed (at least in contract of ThingBuilder.setFeature(String featureId, FeatureDefinition featureDefinition, FeatureProperties featureProperties)`

The implementation however supports nullable `FeatureDefinition` and `FeatureProperties` .. 
So could you:
* change the signature of ThingBuilder.FromScratch.setFeature(...)` and make `FeatureDefinition` and `FeatureProperties` `@Nullable`?
* secondly, pass in `null` if the feature has no definition instead of `.get()` (which will throw an exception if optional was empty)